cmaes.lua

```diff
@@ -0,0 +1,273 @@
+    local best = BestSolution.new(nil,nil,counteval)
+    local iteration = 0 -- iteration of the optimize loop
+
+
+    local function ask()
+      --[[return a list of lambda candidate solutions according to 
+	    m + sig * Normal(0,C) = m + sig * B * D * Normal(0,I)
+	    --]]
+      -- Eigendecomposition: first update B, D and invsqrtC from C
+      -- postpone in case to achieve O(N^2)
+      if counteval - eigeneval > lambda/(c1+cmu)/C:size(1)/10 then
+        eigeneval = counteval
+        C = torch.triu(C) + torch.triu(C,1):t() -- enforce symmetry
+        D, B = torch.symeig(C,'V') -- eigen decomposition, B==normalized eigenvectors, O(N^3)
+        D = torch.sqrt(D)  -- D contains standard deviations now
+        invsqrtC = B * torch.diag(torch.pow(D,-1)) * B:t()
+      end
+      res = torch.Tensor(lambda,D:size(1))
+      for k=1,lambda do --repeat lambda times
+        z = D:clone()
+        z:apply(function(d) return d * torch.normal(0,1) end)
+        res[{k,{}}] = torch.add(xmean, (B * z) * sigma)
+      end
+
+      return res
+    end
+
+
+    --[[update the evolution paths and the distribution parameters m,
+    sigma, and C within CMA-ES.
+    
+    Parameters
+    ----------
+        `arx` 
+            a list of solutions, presumably from `ask()`
+        `fitvals` 
+            the corresponding objective function values --]]
+    local function tell(arx, _fitvals)
+      -- bookkeeping, preparation
+      counteval = counteval + _fitvals:size(1)  -- slightly artificial to do here
+      N = xmean:size(1)             -- convenience short cuts
+      local iN = torch.range(1,N)
+      local xold = xmean:clone()
+
+      -- Sort by fitness and compute weighted mean into xmean
+      fitvals, arindex = torch.sort(_fitvals)
+      arx = arx:index(1, arindex[{{1, mu}}]) -- sorted arx
+
+      table.insert(f_hist, fitvals[1])
+      best:update(arx[1], fitvals[1], counteval)
+
+      xmean:zero()
+      xmean:addmv(arx:t(), weights) --dot product
+
+      -- Cumulation: update evolution paths
+      local y = xmean - xold
+      local z = invsqrtC * y -- == C^(-1/2) * (xnew - xold)
+
+      local c = (cs * (2-cs) * mueff)^0.5 / sigma
+      ps = ps - ps * cs + z * c -- exponential decay on ps
+      local hsig = (torch.sum(torch.pow(ps,2)) / 
+        (1-(1-cs)^(2*counteval/lambda)) / N  < 2 + 4./(N+1))
+      hsig = hsig and 1.0 or 0.0 --use binary numbers
+
+      local c = (cc * (2-cc) * mueff)^0.5 / sigma
+      pc = pc - pc * cc + y * c * hsig -- exponential decay on pc
+
+      -- Adapt covariance matrix C
+      local c1a = c1 - (1-hsig^2) * c1 * cc * (2-cc)
+      -- for a minor adjustment to the variance loss by hsig
+      for i=1,N do
+        for j=1,N do
+          local r = torch.range(1,mu)
+          r:apply(function(k) 
+              return weights[k] * (arx[k][i]-xold[i]) * (arx[k][j]-xold[j]) end)
+          Cmuij = torch.sum(r) / sigma^2  -- rank-mu update
```

Comment:
Use `local Cmuij`.
I'm able to notice the undeclared variables by requiring [https://github.com/deepmind/strict](strict).